File: proposed-projects/exercise-api/scripts/driver.ts
import { readFileSync, writeSync } from "node:fs";
import { pathToFileURL } from "node:url";

// Not a package module: a subprocess entry point run_driver shells out to via
// `pnpm dlx tsx`, loading `<target>/src/index.ts` the way run_vitest_suite's alias
// does. `--adapt` applies that package's default_export rule.

const TYPE_NAMES: Record<string, string> = {
  rulechar: "char",
  char: "char",
  rulecharexclude: "char_exclude",
  charexclude: "char_exclude",
  char_exclude: "char_exclude",
  ruleend: "end",
  end: "end",
};

type Result = {
  ok: boolean;
  error_type: string | null;
  error_pos: number | null;
  rules: unknown[] | null;
  elapsed_ns: number | null;
  construct_ns: number | null;
  add_ns: number | null;
};

const LOAD_ERROR = { ok: false, error_type: "LoadError", error_pos: null, rules: null, elapsed_ns: null };

function parseArg(name: string): string {
  const index = process.argv.indexOf(`--${name}`);
  if (index === -1 || index + 1 >= process.argv.length) {
    throw new Error(`missing required argument --${name}`);
  }
  return process.argv[index + 1];
}

function parseCount(name: string, fallback: number): number {
  const index = process.argv.indexOf(`--${name}`);
  return index === -1 ? fallback : Number(process.argv[index + 1]);
}

function normalizeType(raw: unknown): string {
  const name = String(raw).toLowerCase().replace(/-/g, "_").replace(/^ruletype\./, "");
  return TYPE_NAMES[name] ?? TYPE_NAMES[name.replace(/_/g, "")] ?? name;
}

function normalizeRule(rule: any): Record<string, unknown> {
  const rawType = rule?.type ?? rule?.constructor?.name;
  const normalized: Record<string, unknown> = { type: normalizeType(rawType) };
  if (rule !== null && typeof rule === "object" && "value" in rule) {
    normalized.value = rule.value;
  }
  return normalized;
}

function runCase(GBNF: any, grammar: string, input: string): Result {
  const start = process.hrtime.bigint();
  try {
    const parser = GBNF(grammar);
    const constructed = process.hrtime.bigint();
    const state = parser.add(input);
    const added = process.hrtime.bigint();
    const rules = Array.from(state as Iterable<unknown>, normalizeRule);
    return {
      ok: true,
      error_type: null,
      error_pos: null,
      rules,
      elapsed_ns: Number(process.hrtime.bigint() - start),
      construct_ns: Number(constructed - start),
      add_ns: Number(added - constructed),
    };
  } catch (e: any) {
    const pos = e?.pos;
    return {
      ok: false,
      error_type: e?.constructor?.name ?? e?.name ?? String(e),
      error_pos: typeof pos === "number" ? pos : null,
      rules: null,
      elapsed_ns: Number(process.hrtime.bigint() - start),
      construct_ns: null,
      add_ns: null,
    };
  }
}

async function load(target: string, adapt: boolean): Promise<any> {
  const mod = await import(pathToFileURL(`${target}/src/index.ts`).href);
  let GBNF = mod.default;
  if (adapt && GBNF === undefined && "GBNF" in mod) {
    GBNF = mod.GBNF;
  }
  if (typeof GBNF !== "function") {
    throw new Error("no GBNF entry point");
  }
  return GBNF;
}

function write(record: unknown): void {
  writeSync(1, JSON.stringify(record, (_, v) => (typeof v === "bigint" ? Number(v) : v)) + "\n");
}

async function main(): Promise<void> {
  const target = parseArg("target");
  const adapt = process.argv.includes("--adapt");
  const repeat = parseCount("repeat", 1);
  const warmup = parseCount("warmup", 0);
  console.log = console.error;
  const cases = readFileSync(0, "utf-8")
    .split("\n")
    .filter((line) => line.trim())
    .map((line) => JSON.parse(line));

  let GBNF: any = null;
  try {
    GBNF = await load(target, adapt);
  } catch {
    GBNF = null;
  }
  if (GBNF === null) {
    for (const _ of cases) write(LOAD_ERROR);
    return;
  }

  // A case may carry its own budget: the timeout covers the whole list, so without one
  // the slowest case dictates how often every other case can be measured.
  const budgets = cases.map((c) => ({
    warmup: c.warmup ?? warmup,
    repeat: Math.max(c.repeat ?? repeat, 1),
  }));
  const samples = cases.map(() => ({ construct_ns: [] as (number | null)[], add_ns: [] as (number | null)[] }));
  const results: Record<string, unknown>[] = cases.map(() => ({}));
  const passes = Math.max(0, ...budgets.map((b) => b.warmup + b.repeat));
  for (let pass = 0; pass < passes; pass++) {
    for (let index = 0; index < cases.length; index++) {
      if (pass >= budgets[index].warmup + budgets[index].repeat) continue;
      const { construct_ns, add_ns, ...rest } = runCase(GBNF, cases[index].grammar, cases[index].input);
      if (pass >= budgets[index].warmup) {
        samples[index].construct_ns.push(construct_ns);
        samples[index].add_ns.push(add_ns);
      }
      results[index] = rest;
    }
  }
  for (let index = 0; index < cases.length; index++) {
    const keep = repeat > 1 || "repeat" in cases[index];
    write(keep ? { ...results[index], ...samples[index] } : results[index]);
  }
}

main();

File: proposed-projects/exercise-api/scripts/driver.py
import argparse
import importlib
import importlib.util
import json
import os
import sys
import time
from pathlib import Path

# Not a package module: a subprocess entry point run_driver shells out to with the
# port root on sys.path, mirroring execute-test-suite's PYTHONPATH strategy. Must
# stay stdlib-only; it runs inside whatever interpreter loads the port.

PHASES = ("construct_ns", "add_ns")
LOAD_ERROR = {"ok": False, "error_type": "LoadError", "error_pos": None, "rules": None, "elapsed_ns": None}
TYPE_NAMES = {
    "rulechar": "char",
    "char": "char",
    "rulecharexclude": "char_exclude",
    "charexclude": "char_exclude",
    "char_exclude": "char_exclude",
    "ruleend": "end",
    "end": "end",
}


def load(target: Path, adapt: bool):
    sys.path.insert(0, str(target))
    if adapt and (target / "gbnf.py").is_file() and not (target / "gbnf").is_dir():
        spec = importlib.util.spec_from_file_location(
            "gbnf", target / "__init__.py", submodule_search_locations=[str(target)]
        )
        package = importlib.util.module_from_spec(spec)
        sys.modules["gbnf"] = package
        spec.loader.exec_module(package)
    return importlib.import_module("gbnf").GBNF


def normalize_type(raw) -> str:
    raw = getattr(raw, "value", raw)
    name = str(raw).lower().replace("-", "_").removeprefix("ruletype.")
    return TYPE_NAMES.get(name) or TYPE_NAMES.get(name.replace("_", "")) or name


def normalize_value(value):
    if isinstance(value, (list, tuple)):
        return [normalize_value(v) for v in value]
    if isinstance(value, dict):
        return {str(k): normalize_value(v) for k, v in value.items()}
    return value


def normalize_rule(rule) -> dict:
    if isinstance(rule, dict):
        raw_type, has_value, value = rule.get("type"), "value" in rule, rule.get("value")
    else:
        raw_type = getattr(rule, "type", None) or type(rule).__name__
        has_value, value = hasattr(rule, "value"), getattr(rule, "value", None)
    normalized = {"type": normalize_type(raw_type)}
    if has_value:
        normalized["value"] = normalize_value(value)
    return normalized


def run_case(GBNF, case: dict) -> dict:
    start = time.perf_counter_ns()
    try:
        grammar = GBNF(case["grammar"])
        constructed = time.perf_counter_ns()
        state = grammar.add(case["input"])
        added = time.perf_counter_ns()
        rules = [normalize_rule(rule) for rule in state]
        return {
            "ok": True,
            "error_type": None,
            "error_pos": None,
            "rules": rules,
            "elapsed_ns": time.perf_counter_ns() - start,
            "construct_ns": constructed - start,
            "add_ns": added - constructed,
        }
    except Exception as e:
        pos = getattr(e, "pos", None)
        return {
            "ok": False,
            "error_type": type(e).__name__,
            "error_pos": pos if isinstance(pos, int) and not isinstance(pos, bool) else None,
            "rules": None,
            "elapsed_ns": time.perf_counter_ns() - start,
            "construct_ns": None,
            "add_ns": None,
        }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--target", type=Path, required=True)
    parser.add_argument("--adapt", action="store_true")
    parser.add_argument("--repeat", type=int, default=1)
    parser.add_argument("--warmup", type=int, default=0)
    args = parser.parse_args()

    out = os.fdopen(os.dup(1), "w")
    sys.stdout = sys.stderr
    cases = [json.loads(line) for line in sys.stdin if line.strip()]

    try:
        GBNF = load(args.target, args.adapt)
    except Exception:
        GBNF = None
    if GBNF is None:
        for _ in cases:
            out.write(json.dumps(LOAD_ERROR) + "\n")
            out.flush()
        return

    # A case may carry its own budget: the timeout covers the whole list, so without one
    # the slowest case dictates how often every other case can be measured.
    budgets = [(case.get("warmup", args.warmup), max(case.get("repeat", args.repeat), 1)) for case in cases]
    samples = [{phase: [] for phase in PHASES} for _ in cases]
    results: list[dict] = [{} for _ in cases]
    for index in range(max((warmup + repeat for warmup, repeat in budgets), default=0)):
        for position, (case, (warmup, repeat)) in enumerate(zip(cases, budgets)):
            if index >= warmup + repeat:
                continue
            result = run_case(GBNF, case)
            phases = {phase: result.pop(phase) for phase in PHASES}
            if index >= warmup:
                for phase, value in phases.items():
                    samples[position][phase].append(value)
            results[position] = result
    for case, sample, result in zip(cases, samples, results):
        keep = args.repeat > 1 or "repeat" in case
        out.write(json.dumps({**result, **sample} if keep else result, default=repr) + "\n")
    out.flush()


if __name__ == "__main__":
    main()
